### ISSUE TITLE
Look into isort and mypy for backend environment

### ISSUE THREAD

**Issue by andrewtavis** [activist-org/activist#329]
### Terms

- [X] I have searched [open and closed feature requests](https://github.com/activist-org/activist/issues?q=is%3Aissue+label%3Afeature)
- [X] I agree to follow activist's [Code of Conduct](https://github.com/activist-org/activist/blob/main/.github/CODE_OF_CONDUCT.md)

### Description

It would be good to check the packages [isort](https://github.com/PyCQA/isort) and [mypy](https://github.com/python/mypy) to see if we should add them to the backend dependencies. Maybe `black` will already cover `isort`, but adding in `mypy` for static typing would likely be very worthwhile 😊

### Contribution

Happy to discuss this with people and either help implementation or do it myself!

**Comment by to-sta** [comment]
I am interested in working on the mypy portion of this issue 😃.

**Comment by andrewtavis** [comment]
Me too, @to-sta 😝😝😅😊 But you can take it :) Can you just confirm that with `black` we don’t need `isort`?

**Comment by to-sta** [comment]
`black` does have an overlap with `isort` but does not replace the use of `isort`. Additionally i would also recommend the use of `flake8` or [ruff](https://github.com/astral-sh/ruff ) and to configure all 3 in a pyproject.toml file for example.

I think it would make sense to seperate mypy and isort (potentially black, ruff) into another issue. Since they can be worked on independently of each other.

If you want to work on mypy then i could switch to working on isort, black and ruff 😃.

In the end they all come together in the backend GitHub workflow 😃.

**Comment by andrewtavis** [comment]
Would `mypy` also have a workflow run to check for it? I mostly just want to try to implement the types Python code :) We can focus on `mypy` in this one and then you could make another issue for the others? 😊

**Comment by andrewtavis** [comment]
I’m also fine with ruff over flake8, btw :)

**Comment by to-sta** [comment]
Yes, we can just add `mypy` to the backend GitHub workflow and instead of looking for a pre-defined GitHub action for it we can install via pip and run it in the workflow.

Then i will start working on `mypy` and create another issue for `black`, `isort` and `ruff` afterwards.

**Comment by andrewtavis** [comment]
Would you say this issue is good to close, @to-sta? Or are we waiting to see how it works? Also, do you want to talk a bit at the dev sync about the recent changes? I'd be happy to update the schedule with a section for you to go over some of the updates 😊

**Comment by to-sta** [comment]
Sure, i am happy to talk about the changes in the next dev sync 😃. 

I think it makes sense to keep the issue open since there are a lot of options for configuring `mypy` and there are good reasons for ignoring some of them. I think you mentioned you want to try it out and that is possible now. 
I would suggest running mypy and then discussing the errors and configuring `mypy`. 
After that we need to fix errors that occured. For example currently we don't have type hints for the methods in models.py:

```python
class SupportEntityType(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    name = models.CharField(max_length=255)

    def __str__(self):
        return self.name

# with type hints
    def __str__(self) -> str:
        return self.name

```
And if you are happy with the results then we could implement a GitHub workflow for it.

**Comment by andrewtavis** [comment]
Sounds good, @to-sta! 😊

**Comment by to-sta** [comment]
@andrewtavis @momanyisamuel 

I ran mypy with the current configuration and at the moment we are having 44 errors. Mostly missing type annotations (39/44) but also missing type parameters for generic type "*"(5/44). 

Solving missing type annotations is straightforward and has the benefit of improving readablitiy and code highlighting. 

```python
    def __str__(self):
        return self.name

    # with type annotations
    def __str__(self) -> str:
        return self.name
```
Missing generic type on the other hand is not as straight forward in my opinion. What you think @momanyisamuel? Might slow down progress in my opinion and i think this also needs to be address in the styleguide if we want to enforce this.
There are two location where we are having these errors serializer.py and admin.py.

serializer.py
```python
class UserSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        ....

# with specified type
class UserSerializer(serializers.ModelSerializer[User]):
    class Meta:
        model = User
        ....
```

admin.py
```python
class UserAdmin(admin.ModelAdmin):
    list_display = ("username", "first_name", "last_name", "email", "creation_date")
# with specified type
class UserAdmin(admin.ModelAdmin[User]):
    list_display = ("username", "first_name", "last_name", "email", "creation_date")
```

Additional to get ModelAdmin to work properly we have to add two lines of code to settings.py

```python
import django_stubs_ext
django_stubs_ext.monkeypatch()
```
We could either move forward with this or ignore these error with `disallow_any_generics = False`.


<details>
<summary>Here are all the errors</summary>

```bash
events\models.py:45: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:57: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:70: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:85: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:93: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:101: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:109: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:117: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:125: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:44: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:52: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:67: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:75: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:86: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:94: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:110: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:119: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:127: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:135: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:144: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:152: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:160: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:35: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:48: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:61: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:69: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:77: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:29: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:43: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:64: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:72: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:80: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:88: error: Function is missing a type annotation  [no-untyped-def]
entities\admin.py:6: error: Missing type parameters for generic type "ModelAdmin"  [type-arg]
authentication\admin.py:6: error: Missing type parameters for generic type "ModelAdmin"  [type-arg]
manage.py:7: error: Function is missing a return type annotation  [no-untyped-def]
manage.py:7: note: Use "-> None" if function does not return a value
manage.py:22: error: Call to untyped function "main" in typed context  [no-untyped-call]
events\serializers.py:6: error: Missing type parameters for generic type "ModelSerializer"  [type-arg]
entities\serializers.py:6: error: Missing type parameters for generic type "ModelSerializer"  [type-arg]
authentication\serializers.py:6: error: Missing type parameters for generic type "ModelSerializer"  [type-arg]
entities\views.py:14: error: Function is missing a type annotation  [no-untyped-def]
entities\views.py:26: error: Function is missing a type annotation  [no-untyped-def]
entities\views.py:35: error: Function is missing a type annotation  [no-untyped-def]
entities\views.py:45: error: Function is missing a type annotation  [no-untyped-def]
Found 44 errors in 11 files (checked 42 source files)
```
</details>

**Comment by momanyisamuel** [comment]
@to-sta i think it would be great to implement this, no need to worry about it slowing things down, I can assist with adding this to the style-guide so that it is enforced as you suggested.

**Comment by to-sta** [comment]
Ok, good @momanyisamuel 😃. 

Then we have 3 tasks to wrap this up:

1. Adding type annotations and adjusting settings.py
2. Adding mypy to the backend workflow
3. Updating the styleguide 

Would be great if you could update the styleguide :).

**Comment by andrewtavis** [comment]
Thanks to you both! I’ll be back to more fully check things in a few days :)

**Comment by andrewtavis** [comment]
For this we now just need to update the styleguide, right? :) Just checking in 😊

**Comment by to-sta** [comment]
@andrewtavis I have not added mypy to the workflow yet and we need to update the styleguide. We should mention the changes including ruff and isort in the styleguide.

**Comment by andrewtavis** [comment]
Thanks for the update, @to-sta! It’s getting my hard to keep track of everything, which I guess it a good thing 😅

**Comment by andrewtavis** [comment]
Let's chat in the dev sync today about what all else is needed for this, @to-sta :)

**Comment by andrewtavis** [comment]
Closed via #430 :) Thanks @to-sta!
